### ISSUE TITLE
Weird term name behaviour

### ISSUE THREAD

**Issue by StaccatoSemibreve** [unisonweb/unison#3536]
In my library, I have a namespace `src.util` which contains various functions, and a subnamespace `src.util.ops` which contains operator aliases for many of them.

In a doc table at `readme.cheatsheets.long.standard` that lists many of the function names (all of which are organised in this way), it outputs as:

```
Standard (Long)   src.carriers.MapOf   GetOf                FoldOf               Get0Of                Fold1Of               MatchOf
  Mapping           util.mapOf           getMapOf             foldOf               fold0Of               fold1Of               matchMapOf
  Setting           util.setOf           ops.^-               ops.^--              ops.^^-               ops.^^--              ^!-
  Modifying         modifyOf             gettingMapOf         foldingOf            folding0Of            folding1Of            matchingMapOf
  Assigning         assignOf             src.util.gettingOf   src.util.listingOf   src.util.getting0Of   src.util.listing1Of   matchingOf
```

The expected behaviour is that `ops.^-` would be `getOf`, `ops.^--` would be `listOf`, etc.

Output of `names getOf`:
```
names getOf

  Term
  Hash:   #ql7tu27o3f
  Names:  src.util.getOf src.util.ops.^-
  
  Tip: Use `names.global` to see more results.
```

`getOf` is also used in the definition of `gettingOf`, and appears in the expected way there:
```
view gettingOf

  src.util.gettingOf :
    (GetOf r {Ix is, Err js, e} r b
    ->{g,
    Prism1 (GetOf r),
    Lens (GetOf r),
    MLens0 (GetOf r),
    Indexed (GetOf r),
    MLens (GetOf r),
    Getter (GetOf r),
    MLens1 (GetOf r)} GetOf r {Ix (), Err Void, e} s t)
    ->{e, g, Store s} r
  src.util.gettingOf op = Store.get |> getOf op
```

it's also a little odd that `mapOf`, `gettingOf` etc are listed with namespace prefixes, given that they are the only examples of such names in the entire library namespace, though other functions by those names exist in entirely different namespaces in the codebase. Maybe this is somehow connected?

**Comment by aryairani** [comment]
I think the next steps on this are for us to document the desired behavior for the current implementation, and then make a pass to try to make sure that all relevant sites use that same behavior.

**Comment by ceedubs** [comment]
First step is to make a minimal(ish) transcript that tries to reproduce unexpected behavior.

**Comment by pchiusano** [comment]
The logic used to be that the shortest unique suffix of the (fully qualified name with the fewest number of segments) was preferred, falling back to alphabetical if there's a tie. 

So `optics.ops.&&~` would never be selected over `optics.blah`, even though `&&~` might be a unique suffix.

Apparently that got changed though?

**Comment by ChrisPenner** [comment]
For anyone interested in digging in further, here's the code which sorts names by priority:

https://github.com/unisonweb/unison/blob/trunk/parser-typechecker/src/Unison/PrettyPrintEnv/Names.hs#L31-L41

But call-sites can optionally bias towards  a given namespace prefix using `biasTo`:
https://github.com/unisonweb/unison/blob/trunk/parser-typechecker/src/Unison/PrettyPrintEnv.hs#L140-L154

In order to identify the issue and find a proper fix it'd be helpful to have a change to the name selection transcript which demonstrates the issue; the transcript is here:
https://github.com/unisonweb/unison/blob/cp/accept-referents/unison-src/transcripts/name-selection.md

**Comment by ChrisPenner** [comment]
@pchiusano 

Yeah, the new behaviour would be, given these names:

* `optics.ops.^. (#123)`
* `optics.getOf (#123)`
* `src.util.getOf (#456)`

If printing (#123) the pretty printer will pick `^.`, since it has the shortest _unique_ suffix, since `gettingOf` would be ambiguous, and `optics.gettingOf` is longer than `^.`.

If we identify that this is unintuitive in a lot of cases we can switch the heuristic, but I'd guess there are probably just as many cases where the opposite behaviour is also unintuitive 😓

**Comment by pchiusano** [comment]
Well, there was some thinking behind the old algorithm, it wasn't chosen at random. :) In fact, we used to pick the name with the shortest unique suffix, falling back to alphabetically, and that gave poor / unpredictable results and was changed. (Like you add a second `map` function to a module and all the sudden it picks the `<$>` alias since that's shorter and more unique)

The "pick the FQN with the smallest number of segments, then take its unique suffix" is easy to explain and reason about, to the point where the user can control what names are selected just by tucking them into some deeper namespace (like putting operator names in `ops`, or putting random aliases for `foo` in `foo.aliases`).
